### PR TITLE
Python 3.x support + PEP8 Cleanup

### PR DESCRIPTION
**Pull request by bsandeepan95**
Kindly review the code and let me know if the commits are okay. I will add the docstrings for functions and methods in a separate pull request as @palday asked me to do earlier.

### CODE REVIEW

File: EDF.py [hunk 1]
```diff
@@ -275,31 +386,39 @@ def readHeader(self):
...
                 meas_info['n_records'] = tot_samps/sum(n_samps)
 
-        self.calibrate = (chan_info['physical_max'] - chan_info['physical_min'])/(chan_info['digital_max'] - chan_info['digital_min']);
-        self.offset    =  chan_info['physical_min'] - self.calibrate * chan_info['digital_min'];
+        self.calibrate = (
+            chan_info['physical_max'] - chan_info['physical_min']
+            ) / (chan_info['digital_max'] - chan_info['digital_min'])
+        self.offset    =  chan_info['physical_min'] - (
+            self.calibrate * chan_info['digital_min'])
+
```
Comment (palday) [inline EDF.py:411]:
```suggestion
        physical_range = chan_info['physical_max'] - chan_info['physical_min']
        digital_range = chan_info['digital_max'] - chan_info['digital_min']
        self.calibrate = physical_range / digital_range
        calibrated_dig_min = self.calibrate * chan_info['digital_min']
        self.offset =  chan_info['physical_min'] - calibrated_dig_min
```

File: EDF.py [hunk 2]
```diff
@@ -308,74 +427,84 @@ def readBlock(self, block):
...
             blocksize = np.sum(chan_info['n_samps']) * meas_info['data_size']
             fid.seek(meas_info['data_offset'] + block * blocksize)
+            
             for i in range(meas_info['nchan']):
                 buf = fid.read(chan_info['n_samps'][i]*meas_info['data_size'])
-                raw = np.asarray(unpack('<{}h'.format(chan_info['n_samps'][i]), buf), dtype=np.float32)
+                
+                raw = np.asarray(
+                    unpack('<{}h'.format(
+                        chan_info['n_samps'][i]), buf), dtype=np.float32)
```
Comment (palday) [inline EDF.py:436]:
```suggestion
                raw = unpack('<{}h'.format(chan_info['n_samps'][i]), buf)
                raw = np.asarray(raw, dtype=np.float32)
```

File: EDF.py [hunk 3]
```diff
@@ -308,74 +427,84 @@ def readBlock(self, block):
...
+            data = np.append(data, self.read_block(block)[channel])
         begsample -= begblock*n_samps
         endsample -= begblock*n_samps
-        return data[begsample:(endsample+1)]
+        return data[begsample:(endsample + 1)]
 
-####################################################################################################
-# the following are a number  of helper functions to make the behaviour of this EDFReader
-# class more similar to https://bitbucket.org/cleemesser/python-edf/
-####################################################################################################
```
Comment (palday) [inline EDF.py:335]: don't delete this

File: EDF.py [hunk 4]
```diff
@@ -308,74 +427,84 @@ def readBlock(self, block):
...
-    print file_in.readSamples(0, 0, 128)
-
-
-if False:
+    def read_signal(self, chanindx):
+        begsample = 0
+        endsample = (
+            self.chan_info['n_samps'][chanindx] * self.meas_info['n_records']
+            ) - 1
+        return self.read_samples(chanindx, begsample, endsample)
```
Comment (palday) [inline EDF.py:473]:
```suggestion
    def read_signal(self, chidx):
        beg = 0
        end = self.chan_info['n_samps'][chidx] * self.meas_info['n_records']
        return self.read_samples(chidx, beg, end - 1)
```

File: README.md [hunk 5]
```diff
@@ -1,10 +1,12 @@
...
+
+The edf file, included for testing, is taken from [EDFBrowser Test Files](https://www.teuniz.net/edf_bdf_testfiles/index.html).
 
-Since EDF is such a widely used format, there exist multiple implementations for reading and writing EDF files. Most Python implementations consist of wrappers around the C-code implementation, which makes installation more cumbersome and reduces portability. This implementation is in pure Python with limited dependencies on external packages.
+The EDF standard was introduced in [1992](https://doi.org/10.1016/0013-4694(92)90009-7). The extension of EDF with annotations was first described in [1998](https://doi.org/10.1016/S0013-4694(98)00029-7) and more formalized with the EDF+ standard that was published in [2003](https://doi.org/10.1016/S1388-2457(03)00123-8). The [EDF](http://www.edfplus.info) website describes both standards and discusses implementation details.
```
Comment (palday) [inline README.md:9]: Why change this?

File: EDF.py [hunk 6]
```diff
@@ -1,3 +1,36 @@
...
+        chan_info should have the following: [
+            'physical_min', 'transducers', 'physical_max', 
+            'digital_max', 'ch_names', 'n_samps', 'units', 'digital_min']
+    
+    The EDF standard was introduced in 1992. The extension of EDF with 
+    annotations was first described in 1998 and more formalized with the EDF+ 
+    standard that was published in 2003. To learn more about both standards and 
+    implementation details, check out https://www.edfplus.info/index.html
+"""
+__version__ = "0.4.0"
```
Comment (palday) [inline EDF.py:26]: Where did the version number come from?

File: EDF.py [hunk 7]
```diff
@@ -1,3 +1,36 @@
...
+__author__ = "https://github.com/robertoostenveld | https://github.com/bsandeepan95>"
```
Comment (palday) [inline EDF.py:27]: I think traditionally this is name + email address. Note also the distinction between author, credits/contributora, and maintainera.

**Comment by palday** [comment]
One thing I forgot: `write_byte` really should be `write_bytes` because (1) it's usually more than 1 byte of data being written and (2) it's a helper function to write things of the `bytes` type.

File: EDF.py [hunk 8]
```diff
@@ -8,32 +41,44 @@
...
-    else:
-        # trim the input to the specified length
-        return buf[0:num]
-
-####################################################################################################
-# the EDF header is represented as a tuple of (meas_info, chan_info)
-# meas_info should have ['record_length', 'magic', 'hour', 'subject_id', 'recording_id', 'n_records', 'month', 'subtype', 'second', 'nchan', 'data_size', 'data_offset', 'lowpass', 'year', 'highpass', 'day', 'minute']
-# chan_info should have ['physical_min', 'transducers', 'physical_max', 'digital_max', 'ch_names', 'n_samps', 'units', 'digital_min']
-####################################################################################################
+    return (str(buf) + ' ' * num) if (num >= 0) else buf[0:num]
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:47]: This is just a short-hand expression of the multi-line if-else condition. The explanation of the code is written in the documentation of the function, which you earlier advised to put in separate pull request. Hence it was removed here. You can open the earlier commit and compare the code with the documentation if required.
fun_padtrim_doc

I expected these confusions to happen. I should have explained earlier why it would be better to consider both the documentation and code edits to review side by side. This is my bad. Let me know if I can do anything to alleviate this.

[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:47]: This syntax is globally accepted and in my system, it has not break in both py2.7 and py3.x yet. You may test it individually and let me know if it does.

File: EDF.py [hunk 9]
```diff
@@ -44,105 +89,155 @@ def close(self):
...
+                # skip this part
+                fid1.read(8)
+                # but write this instead
+                writebyte(fid2, padtrim(str(self.n_records), 8))
+
+                writebyte(fid2, fid1.read(
+                    meas_info['data_offset'] - 236 - 8))
+                
+                blocksize = np.sum(
+                    chan_info['n_samps']) * meas_info['data_size']
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:112]: Truth be told, looking at how hard I went after 80-char limit, I am laughing at myself now 10 months after. Should I click "Commit suggestion" or edit in my code and request pull again?

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1,6 @@
+# Exclude these dirs
+/_pycache_/
+/.vscode/
+
+# files
+EDF Docs.py
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply .gitignore:6]: I wished to remove gitignore from the commit actually.

File: EDF.py [hunk 11]
```diff
@@ -8,32 +41,44 @@
...
         self.fname = None
         self.meas_info = None
         self.chan_info = None
         self.calibrate = None
         self.offset    = None
         self.n_records = 0
-        if fname:
-            self.open(fname)
-
+
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:81]: The function init_properties() contains a few value reset lines. These lines are actually called again in present line 118 when we are closing the EDF file. This function wrap only reduces the redundant lines and now the original call is covered in __init__()

I am explaining this after 10 months of original edit. If it sounds mistaken, let me know.

File: EDF.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (palday) [reply EDF.py:47]: I  understand the Python syntax. I don't understand the need to change to a different syntax. Changes just to make changes aren't necessary.

File: EDF.py [hunk 12]
```diff
@@ -44,105 +89,155 @@ def close(self):
...
+        
         os.remove(tempname)
-        self.fname = None
-        self.meas_info = None
-        self.chan_info = None
-        self.calibrate = None
-        self.offset    = None
-        self.n_records = 0
-        return
+        self.init_properties()
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:118]: The original lines did not handle fname in close(). Hence I wrapped the repetitive lines like this. I wanted to rename the function properly but failed to find a better name. Afterwards, I got busy in office works. Do you have any suggestions?

File: EDF.py (hunk 3 above)
Comment (bsandeepan95) [reply EDF.py:335]: This section is covered in details in the added documentation -
`        Helper Methods: The following are a number of helper functions to make 
                        the behaviour of this EDFReader class more similar to 
                        https://bitbucket.org/cleemesser/python-edf/
            get_signal_text_labels: Convert Signal Text Labels from unicode 
                                    to strings.
            get_n_signals:      Get Number of Channels.
            get_signal_freqs:   Get Signal Frequencies.
            get_n_samples:      Get Total Number of Samples.
            read_signal:        Reads Entire Signal Record and returns the 
                                values as an array.
`

File: README.md (hunk 5 above)
Comment (bsandeepan95) [reply README.md:9]: It's a simple order matter. The purpose of the library is put ahead of information about EDF, as we may safely assume that people looking for edf libraries already know about edf.

File: EDF.py (hunk 7 above)
Comment (bsandeepan95) [reply EDF.py:27]: This one I wanted a discussion about. I don't see any versioning in original library. SO I gave it one by  counting the git commits in original one. If the library is to be without versioning, then I think removal is better.

File: EDF.py (hunk 6 above)
Comment (bsandeepan95) [reply EDF.py:26]: This one I wanted a discussion about. I don't see any versioning in original library. SO I gave it one by  counting the git commits in original one. If the library is to be without versioning, then I think removal is better.

File: EDF.py (hunk 7 above)
Comment (bsandeepan95) [reply EDF.py:27]: Yes. I don't know robert's mail id so I put it like this. Even in original code, he has not added his email id or anything as far as I have checked. I wished to do it properly using his name as author and ours as contributors but I don't have the necessary resources. Hence I had to wait till this code review.

File: EDF.py (hunk 4 above)
Comment (bsandeepan95) [reply EDF.py:473]: won't removing the sample word from the variables make it less dev-friendly? Although the variables will be defined properly in the documentation.

File: EDF.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:118]: @palday @robertoostenveld I just thought about renaming the function as initial_prop() or initial_state_props (influenced from ReactJS). This way, we can address that the function is supposed to set a few object properties to their initial state, whenever invoked. This, i hope, will end any probable confusion of the function purpose for construction and reset of the objects.

Also, the function is created as a method for each object, causing repetition. What if we create it as a global function, or as under an object that only hosts this property which can be inherited by both edfreader and edfwriter? This may seem unnecessary to you and please let me know if so. I am just having ideas.

File: EDF.py [hunk 13]
```diff
@@ -44,105 +89,155 @@ def close(self):
...
             if not 'recording_id' in meas_info:
                 meas_info['recording_id'] = ''
+            
             if not 'subtype' in meas_info:
                 meas_info['subtype'] = 'edf'
+            
             nchan = meas_info['nchan']
-            if not 'ch_names' in chan_info or len(chan_info['ch_names'])<nchan:
+            
+            if ((not 'ch_names' in chan_info) or (
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:141]: I support this change as well. Will edit the code accordingly.

**Comment by bsandeepan95** [comment]
@palday I am new to this pull requests so I there are some changes that I wish to do right away and commit but it seems there are many such ones and github is providing me a "Commit Change" button. If I use that button, will it make numerous new comments for each change I save, or will it stage them together later when the pull is successful? Seeking your guidance.

File: README.md (hunk 5 above)
Comment (palday) [reply README.md:9]: That assumption holds less often than you might otherwise think (as I know from my other job -- people look for EDF libraries because they don't know anything about EDF but suddenly have to work with it). But as a guiding principle, don't change things unnecessarily: https://yihui.org/en/2018/11/cosmetic-changes/

File: EDF.py (hunk 4 above)
Comment (palday) [reply EDF.py:473]: It's a tradeoff with line breaks. Logical lines spread over several physical lines are hard to read. `beg` and `end` indicate beginning and end, while the samples bit is clear from the function call `read_samples`.

File: EDF.py (hunk 3 above)
Comment (palday) [reply EDF.py:335]: There is value to having the documentation close to the relevant code.

File: EDF.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (palday) [reply EDF.py:112]: `commit suggestion` will create a commit with exactly those changes. You'll have to use `git pull` on your local computer to update to the changes made on GitHub directly.

File: EDF.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (palday) [reply EDF.py:47]: This is one of the downsides to runtime typing. The syntax is valid, but you might still get an error depending on the types passed.

[the remark this answers is not in the document]
Comment (palday) [reply EDF.py:47]: > Do you not think that this is somewhat better than a multi-line if-else?

No, I don't. If for no other reason than the order is inverted compared to most other programming constructs (the test occurs between values, which is weird in programming languages, see e.g. the tertiary operator in C/C++/Java ` test ? val_if_true : val_if_false` or R's `ifelse(test, val_if_true, val_if_false`.

> The backstory for this is that when I was making these formatting I was working with embedded engineers who are too paranoid about saving space. I picked up the habit because I had to write similar codes.

The length of the source does not matter. The length of the compiled code does. Source code is for humans. Parsed/compiled code is for computers.

File: EDF.py (hunk 6 above)
Comment (palday) [reply EDF.py:26]: I would just go with `0.2.0` (allowing the previous version to be an implicit `0.1.0`) and follow [SemVer](https://semver.org/).

File: EDF.py (hunk 7 above)
Comment (palday) [reply EDF.py:27]: If you look at the output of `git log` locally, you'll see committer email addresses. :smile:

**Comment by palday** [comment]
I think "Commit change" creates a new commit for every single change, but lots of small commits are fine -- especially because they'll all be squashed during the merge.

Thanks for doing this: I know it's a lot of effort, and we do appreciate this. My review is critical, but it's meant to be constructive.

File: EDF.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:47]: > the order is inverted compared to most other programming constructs
Yes, that is just the philosophy of python . It's meant to be simpler like `val_if_true if test is true else val_if_false`. Though the syntax is not a ternary operator. It is simply an one-line if-else expression such as lambda function expressions, added back in Python 2.5.

https://mail.python.org/pipermail/python-dev/2005-September/056846.html

[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:47]: I just checked the code and errors are unlikely. See,

`padtrim()` is a helper function used to standardize header content to pass into `writebyte()`. In Python 2.x, strings and bytes are synonymous unlike in python 3.x. If you check the logic of `writebyte()`, it has error-handling support which will deal with both bytes and string formats i.e. output will be a bytes value. If some other format is returned (such as float or boolean), which is very unlikely logically, then we have to worry. I am not sure how I should write a unit test for least likely scenario

[the remark this answers is not in the document]
Comment (palday) [reply EDF.py:47]: Again, I'm very familiar with both the syntax and its history in Python. (The same holds for all of my other points: I don't require an introduction to Python, nor an explanation of the differences between Python 2 and Python 3.) Note that I did not call it a tertiary operator, but compared it the tertiary operator in other languages.

The philosophy of Python not to invert order. In the Zen of Python, it's quite clear that there should be a preferred obvious way to do something. The multiline `if` statement is obvious and doesn't involve an unnecessary change. Notice in the link you provide that the former BDFL even mentions a "long discussion" for adding the single-line `if` expression. The long discussion was because simply adding that expression into the language potentially violated the Zen:
> Readability counts. 
....
> There should be one-- and preferably only one --obvious way to do it.

There are two ways here in Python, but the multiline statement is the older and more obvious route.

I'll concede that other programmers find the one-liner easy/ier to read. But the multiline version is not hard to read *and it doesn't involve a cosmetic change to somebody else's codebase*.

[the remark this answers is not in the document]
Comment (palday) [reply EDF.py:47]: There is an advantageous to having consistent types internally. For example, this could be fixed by having the trim case also convert to `str`.  Alternatively, if doing the preferred way of dealing with `str` in Python 3, we assume that everything internally is `str` and only the actual I/O operations are dealing with `bytes` and those are doing the conversion appropriately (then we don't need the conversion to `str` in the pad case). `writebyte` handles the output bit of that, and I think (but haven't checked) that all of the `file.read` are explicitly converted to `str` (or `int`).

File: EDF.py [hunk 14]
```diff
@@ -8,32 +41,44 @@
...
-
-####################################################################################################
-# the EDF header is represented as a tuple of (meas_info, chan_info)
-# meas_info should have ['record_length', 'magic', 'hour', 'subject_id', 'recording_id', 'n_records', 'month', 'subtype', 'second', 'nchan', 'data_size', 'data_offset', 'lowpass', 'year', 'highpass', 'day', 'minute']
-# chan_info should have ['physical_min', 'transducers', 'physical_max', 'digital_max', 'ch_names', 'n_samps', 'units', 'digital_min']
-####################################################################################################
+    return (str(buf) + ' ' * num) if (num >= 0) else buf[0:num]
+
+
+def writebyte(file, content, encoding='utf8'):
```
Comment (palday) [inline EDF.py:50]:
```suggestion
def writebyte(file, content, encoding='ascii'):
```

`writebyte` should default to using ASCII encoding because that's what the EDF spec [actually requires](https://www.edfplus.info/specs/edf.html).

File: EDF.py [hunk 15]
```diff
@@ -8,32 +41,44 @@
...
+        pass
+    except Exception as e:
+        print(type(e))
+        print(str(e))
+        print(
+            "If you see this message, please go to " + 
+            "https://github.com/bids-standard/pyedf/issues" + 
+            " and open an issue there regarding this. Thank you!")
+    finally:
+        file.write(content)
```
Comment (palday) [inline EDF.py:65]:
```suggestion
    if type(content) != bytes:
        content = content.encode(encoding)
      
    file.write(content)
```

(This works in Python 2.7 and Python 3 -- I haven't tested it in older Python 2, but am not interesting in supporting that!)

Also, note that the `finally` clause here doesn't make much sense: if there has been a conversion error, then you don't want to try to write the bytes out to file. It would be better to simply leave out the extra `except Exception as e` and `finally` clauses and have `file.write` be after the `try` block. But in this case, there's an even easier solution (see suggestion).

File: EDF.py [hunk 16]
```diff
@@ -44,105 +89,155 @@ def close(self):
...
+            writebyte(fid, (padtrim(meas_info['recording_id'], 80)))
+
+            writebyte(fid, (padtrim('{:0>2d}.{:0>2d}.{:0>2d}'.format(
+                meas_info['day'], meas_info['month'], 
+                meas_info['year']), 8)))
+            
+            writebyte(fid, (padtrim('{:0>2d}.{:0>2d}.{:0>2d}'.format(
+                meas_info['hour'], meas_info['minute'], 
+                meas_info['second']), 8)))
+
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:169]: I wanted to keep them same as before in original library -
`fid.write(padtrim('{:0>2d}.{:0>2d}.{:0>2d}'.format(meas_info['day'], meas_info['month'], meas_info['year']), 8))`
`fid.write(padtrim('{:0>2d}.{:0>2d}.{:0>2d}'.format(meas_info['hour'], meas_info['minute'], meas_info['second']), 8))`

we can certainly use a variable over and over again to make this more readable.

File: EDF.py (hunk 15 above)
Comment (bsandeepan95) [reply EDF.py:65]: Since I implemented Error Handling block here, I used `finally` clause to help people understand the code flow. Since there is only one line left to do after the block, and I know developers who can indent such lines one level inside the `except` clause thinking it as a mistake on our end. I wish to avoid clearing up such confusion in the future if anyone faces this.

I added the `except Exception as e` clause to ensure people know what to do if they face such error. I can always do away with this. This is only there for least expected scenarios.

`bytes` is not a supported type in Python 2.7 so the `if` clause will not work there. However, as we have already added the support in `except TypeError` clause, how about we put the `content = content.encode(encoding)` line in there? This will ensure that encoding is done correctly for Python 2.7.

File: EDF.py (hunk 14 above)
Comment (bsandeepan95) [reply EDF.py:50]: I remember you asking me to do this a year ago. I forgot. Apologies. :sweat_smile:

File: EDF.py (hunk 15 above)
Comment (palday) [reply EDF.py:65]:
```python
Python 2.7.17 (default, Jul 20 2020, 15:37:01) 
[GCC 7.5.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> x = "a"
>>> type(x)
<type 'str'>
>>> type(x) == bytes
True
>>> type(x) == str
True
```

File: EDF.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:47]: I feel we are having a developer's feud about a line of code. 😅 I do not wish to do this, and I understand your point, and your reasoning is correct. Yet I, for some reason, feel rigid about this line. 😂 

Though what you pointed out is correct. I do not want to be remembered as the guy doing _cosmetic changes_ to this codebase and then being like "I shall defend this short braid of hair till death!". That's not me, so I am making the change now. 💔

File: EDF.py (hunk 15 above)
Comment (bsandeepan95) [reply EDF.py:65]: Right. `bytes` is a synonym for `str` in Python 2.7.

I mean, if we are to do `ascii` encoding either way, why not enforce the line in the `TypeError` block, which will be triggered in Python 2.7 since default `str` encoding scheme is `utf-8`.

File: EDF.py (hunk 7 above)
Comment (bsandeepan95) [reply EDF.py:27]: @palday your email id is not there.

File: EDF.py (hunk 15 above)
Comment (bsandeepan95) [reply EDF.py:65]: Okay, I ran a bit of code for testing the error handling block in Python 2.7 and I feel the block is necessary. Run the following example - 

```python
# -*- coding: utf-8 -*-
enc = 'ascii'
x = 'µ Õ ß'

print x

try:
    y = x.encode(enc)
except Exception as e:
    print("Error Message: " + str(e))
    print("Error Type: " + str(type(e)))
    y = 'fail'

print y
print (x == y)
 ```

The output will be like the following - 
```python
Python 2.7.16 (default, Jul 13 2019, 16:01:51)
[GCC 8.3.0] on linux2

µ Õ ß
Error Message: 'ascii' codec can't decode byte 0xc2 in position 0: ordinal not in range(128)
Error Type: <type 'exceptions.UnicodeDecodeError'>
fail
False
```

So errors will occur if there are non-ascii characters, and the code will fail. Now the question is if it is better to let the code fail, or if we should handle the error gracefully and display it to the console for the developer. Let me know your opinion about this @palday .

Comment (bsandeepan95) [reply EDF.py:65]: @palday waiting for your response.

**Comment by bsandeepan95** [comment]
> 
> 
> I think "Commit change" creates a new commit for every single change, but lots of small commits are fine -- especially because they'll all be squashed during the merge.
> 
> Thanks for doing this: I know it's a lot of effort, and we do appreciate this. My review is critical, but it's meant to be constructive.

Please do. I am grateful for your review. It is making me rethink some of my past coding choices and shaping my perspective for future work. I may sometimes come out difficult to communicate, but I am trying to improve my own bugs too here. 😄 

Strict review is necessary. Else we get unexpected drivers issues that disencourage me from using Ubuntu Linux in my AMD Vega Graphics Workstation. >__>

File: EDF.py [hunk 17]
```diff
@@ -44,105 +89,155 @@ def close(self):
...
-            for key in ['physical_min', 'transducers', 'physical_max', 'digital_max', 'ch_names', 'n_samps', 'units', 'digital_min']:
+            for key in [
+                'physical_min', 'transducers', 'physical_max', 'digital_max', 
+                'ch_names', 'n_samps', 'units', 'digital_min']:
                 chan_info[key] = np.asarray(chan_info[key])
 
             for i in range(meas_info['nchan']):
-                fid.write(padtrim(    chan_info['ch_names'][i], 16))
+                writebyte(fid, (
+                    padtrim(    chan_info['ch_names'][i], 16)))
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:186]: I just kept it as the space was in main author's commits. I am not changing the for loops for now.

File: EDF.py [hunk 18]
```diff
@@ -44,105 +89,155 @@ def close(self):
...
         self.meas_info = meas_info
         self.chan_info = chan_info
-        self.calibrate = (chan_info['physical_max'] - chan_info['physical_min'])/(chan_info['digital_max'] - chan_info['digital_min']);
-        self.offset    =  chan_info['physical_min'] - self.calibrate * chan_info['digital_min'];
+        self.calibrate = (
+            chan_info['physical_max'] - chan_info['physical_min']) / (
+                chan_info['digital_max'] - chan_info['digital_min'])
+
+        self.offset    =  chan_info['physical_min'] - (
+            self.calibrate * chan_info['digital_min'])
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:231]: Smooth

File: EDF.py [hunk 19]
```diff
@@ -152,9 +247,11 @@ def writeBlock(self, data):
 
                 assert(len(raw)==chan_info['n_samps'][i])
                 if min(raw)<chan_info['physical_min'][i]:
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:249]: done

File: EDF.py [hunk 20]
```diff
@@ -152,9 +247,11 @@ def writeBlock(self, data):
...
-                    warnings.warn('Value exceeds physical_min: ' + str(min(raw)) );
+                    warnings.warn(
+                        'Value exceeds physical_min: ' + str(min(raw)))
                 if max(raw)>chan_info['physical_max'][i]:
-                    warnings.warn('Value exceeds physical_max: '+ str(max(raw)));
+                    warnings.warn(
+                        'Value exceeds physical_max: '+ str(max(raw)))
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:254]: 👍

File: EDF.py [hunk 21]
```diff
@@ -165,92 +262,105 @@ def writeBlock(self, data):
...
             meas_info['minute'] = minute
             meas_info['second'] = second
-            # date = datetime.datetime(year + 2000, month, day, hour, minute, sec)
-            # meas_info['meas_date'] = calendar.timegm(date.utctimetuple())
+            date = datetime.datetime(
+                year + 2000, month, day, hour, minute, second)
+            meas_info['meas_date'] = calendar.timegm(date.utctimetuple())
 
-            meas_info['data_offset'] = header_nbytes = int(fid.read(8).decode())
+            meas_info['data_offset'] = header_nbyte = int(fid.read(8).decode())
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:313]: That's kept by original author. Only @robertoostenveld  knows what purpose they may serve.

File: EDF.py [hunk 22]
```diff
@@ -165,92 +262,105 @@ def writeBlock(self, data):
...
+            chan_info['units']        = [
+                fid.read(8).strip().decode() for ch in channels]
+            chan_info['physical_min'] = np.array([
+                float(fid.read(8).decode()) for ch in channels])
+            chan_info['physical_max'] = np.array([
+                float(fid.read(8).decode()) for ch in channels])
+            chan_info['digital_min']  = np.array([
+                float(fid.read(8).decode()) for ch in channels])
+            chan_info['digital_max']  = np.array([
+                float(fid.read(8).decode()) for ch in channels])
```
[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:355]: That's only for four lines. Better to keep the repetition as is. Later any developers can comprehend easily.
In one year, I learned that clean code has various meanings.

[the remark this answers is not in the document]
Comment (bsandeepan95) [reply EDF.py:355]: What may have been more helpful is good comments.

File: README.md (hunk 5 above)
Comment (bsandeepan95) [reply README.md]: Editing done. Please review.

**Comment by bsandeepan95** [comment]
@palday, check the new commits

**Comment by palday** [comment]
I need to write some tests to check this, but it's on my radar.

**Comment by bsandeepan95** [comment]
@palday I know we agreed to add the latest documentations in a later PR, but will it help the tests if I add them in this PR?

**Comment by palday** [comment]
Having gone back through the history of this PR .... I'm somewhat surprised that it didn't take advantage of the work already done in #4.  But we can't change the past.

I think the best solution for now is to merge this into a `py3k` branch and provide documentation for installing from that branch for Python 3 support. I'll discuss with @robertoostenveld how to best package up all the work from #4 and here together and hopefully get that back into the trunk branch and set up a PyPI release. 

There seem to be some merge conflicts (which I don't quite understand since `py3k` currently is the same as `master` and `master` didn't have merge conflicts), but I'll have to look at those in a few hours and then I'll merge.

**Comment by bsandeepan95** [comment]
Great! Let me know if I can be of help in merging #4 into the py3k branch. 

@palday , I tried to write the code as a dual support for python 2.7 and 3+. Has there been any edge cases where the tests failed?

Meanwhile, I will add the documentation shortly and make another PR.
